stan/math/rev/mat/fun/matrix_exp_action.hpp

```diff
@@ -0,0 +1,398 @@
+  double t_;
+  double* Ad_;
+  double* Bd_;
+  vari** variRefA_;
+  vari** variRefexpAB_;
+
+  /**
+   * Constructor for matrix_exp_action_vari.
+   *
+   * All memory allocated in
+   * ChainableStack's stack_alloc arena.
+   *
+   * It is critical for the efficiency of this object
+   * that the constructor create new varis that aren't
+   * popped onto the var_stack_, but rather are
+   * popped onto the var_nochain_stack_. This is
+   * controlled to the second argument to
+   * vari's constructor.
+   *
+   * @param A matrix
+   * @param B matrix
+   * @param t double
+   */
+  matrix_exp_action_vari(const Eigen::Matrix<Ta, N, N>& A,
+                         const Eigen::Matrix<double, N, Cb>& B, const double& t)
+      : vari(0.0),
+        n_(A.rows()),
+        B_cols_(B.cols()),
+        A_size_(A.size()),
+        B_size_(B.size()),
+        t_(t),
+        Ad_(ChainableStack::instance().memalloc_.alloc_array<double>(A_size_)),
+        Bd_(ChainableStack::instance().memalloc_.alloc_array<double>(B_size_)),
+        variRefA_(
+            ChainableStack::instance().memalloc_.alloc_array<vari*>(A_size_)),
+        variRefexpAB_(
+            ChainableStack::instance().memalloc_.alloc_array<vari*>(B_size_)) {
+    using Eigen::Map;
+    using Eigen::MatrixXd;
+    for (size_type i = 0; i < A.size(); ++i) {
+      variRefA_[i] = A.coeffRef(i).vi_;
+      Ad_[i] = A.coeffRef(i).val();
+    }
+    for (size_type i = 0; i < B.size(); ++i) {
+      Bd_[i] = B.coeffRef(i);
+    }
+    matrix_exp_action_handler handle;
+    MatrixXd expAB = handle.action(Map<MatrixXd>(Ad_, n_, n_),
+                                   Map<MatrixXd>(Bd_, n_, B_cols_), t_);
+    for (size_type i = 0; i < expAB.size(); ++i)
+      variRefexpAB_[i] = new vari(expAB.coeffRef(i), false);
+  }
+
+  virtual void chain() {
+    using Eigen::Map;
+    using Eigen::MatrixXd;
+    MatrixXd adjexpAB(n_, B_cols_);
+
+    for (size_type i = 0; i < adjexpAB.size(); ++i)
+      adjexpAB(i) = variRefexpAB_[i]->adj_;
+
+    MatrixXd adjA = exp_action_chain_vd<N, Cb>(Ad_, Bd_, adjexpAB, t_);
+
+    for (size_type i = 0; i < A_size_; ++i) {
+      variRefA_[i]->adj_ += adjA(i);
+    }
+  }
+};
+
+/**
+ * Return product of exp(At) and B, where A is a NxN matrix,
+ * B is a NxCb matrix, and t is a double
+ * @tparam Ta scalar type matrix A
+ * @tparam N Rows and cols matrix A, also rows of matrix B
+ * @tparam Tb scalar type matrix B
+ * @tparam Cb Columns matrix B
+ * @param[in] A Matrix
+ * @param[in] B Matrix
+ * @param[in] t double
+ * @return exponential of At multiplies B
+ */
+template <typename Ta, int N, typename Tb, int Cb>
+inline typename boost::enable_if_c<boost::is_same<Ta, var>::value
+                                       || boost::is_same<Tb, var>::value,
+                                   Eigen::Matrix<var, N, Cb> >::type
+matrix_exp_action(const Eigen::Matrix<Ta, N, N>& A,
+                  const Eigen::Matrix<Tb, N, Cb>& B, const double& t = 1.0) {
```

Comment:
I think that for primitive types such as `double`, you don't need to specify `const` and `&`, as the compiler passes by reference automatically.